src/findone-parallel-performance.c

```diff
@@ -0,0 +1,239 @@
+   mongoc_collection_t *coll;
+   int i;
+   bson_t filter = BSON_INITIALIZER;
+
+   coll = mongoc_client_get_collection (ctx->client, "perftest", "coll");
+   bson_append_int32 (&filter, "_id", 3, 0);
+
+   for (i = 0; i < ctx->n_findone_to_run; i++) {
+      mongoc_cursor_t *cursor;
+      const bson_t *doc;
+      bson_error_t error;
+
+      cursor = mongoc_collection_find_with_opts (
+         coll, &filter, NULL /* opts */, NULL /* read_prefs */);
+      /* iterate once to send "find" */
+      if (mongoc_cursor_next (cursor, &doc)) {
+         MONGOC_ERROR (
+            "Error: unexpected document returned from collection: %s",
+            bson_as_json (doc, NULL));
+         abort ();
+      }
+      if (mongoc_cursor_error (cursor, &error)) {
+         MONGOC_ERROR ("Error from cursor iteration: %s", error.message);
+         abort ();
+      }
+      mongoc_cursor_destroy (cursor);
+   }
+
+   mongoc_collection_destroy (coll);
+   bson_destroy (&filter);
+   return NULL;
+}
+
+static void
+findone_parallel_perf_task (perf_test_t *test)
+{
+   findone_parallel_perf_test_t *findone_parallel_test =
+      (findone_parallel_perf_test_t *) test;
+   int i;
+   int ret;
+
+   for (i = 0; i < findone_parallel_test->n_threads; i++) {
+      findone_parallel_thread_context_t *ctx;
+
+      ctx = &findone_parallel_test->contexts[i];
+      ret = pthread_create (
+         &ctx->thread, NULL /* attr */, _findone_parallel_perf_thread, ctx);
+      if (ret != 0) {
+         MONGOC_ERROR ("Error: pthread_create returned %d", ret);
+         abort ();
+      }
+   }
+
+   for (i = 0; i < findone_parallel_test->n_threads; i++) {
+      findone_parallel_thread_context_t *ctx;
+
+      ctx = &findone_parallel_test->contexts[i];
+      ret = pthread_join (ctx->thread, NULL /* out */);
+      if (ret != 0) {
+         MONGOC_ERROR ("Error: pthread_join returned %d", ret);
+         abort ();
+      }
+   }
+}
+
+static void
+findone_parallel_perf_setup (perf_test_t *test)
+{
+   mongoc_uri_t *uri;
+   findone_parallel_perf_test_t *findone_parallel_test =
+      (findone_parallel_perf_test_t *) test;
+   mongoc_client_pool_t *pool;
+   mongoc_client_t *client;
+   mongoc_database_t *db;
+   bson_error_t error;
+
+   uri = mongoc_uri_new (NULL);
+   pool = mongoc_client_pool_new (uri);
+   findone_parallel_test->pool = pool;
+   findone_parallel_test->contexts =
+      (findone_parallel_thread_context_t *) bson_malloc0 (
+         findone_parallel_test->n_threads *
+         sizeof (findone_parallel_thread_context_t));
+
+   client = mongoc_client_pool_pop (findone_parallel_test->pool);
+   db = mongoc_client_get_database (client, "perftest");
+   if (!mongoc_database_drop (db, &error)) {
+      MONGOC_ERROR ("database_drop: %s\n", error.message);
+      abort ();
+   }
+   mongoc_database_destroy (db);
+   mongoc_client_pool_push (pool, client);
+   mongoc_uri_destroy (uri);
+}
+
+static void
+findone_parallel_perf_teardown (perf_test_t *test)
```

Comment:
The test /itself/ is not released, but that's because it's in a statically-sized array. +1.